module_utils/network_lsr/argument_validator.py

```diff
@@ -305,9 +306,10 @@ def _validate_impl(self, value, name):
 
 
 class ArgValidatorList(ArgValidator):
-    def __init__(self, name, nested, default_value=None):
+    def __init__(self, name, nested, enum_values=None, default_value=None):
         ArgValidator.__init__(self, name, required=False, default_value=default_value)
         self.nested = nested
+        self.enum_values = enum_values
```

Comment:
`enum_values` does not seem a good name for a list of regular expressions.

Reply:
the `enum_values` used everywhere in that file, I just want to keep it consistent

Reply:
The name is used for things that are conceptionally similar and for which the name makes sense. In particular, it's a list of names (string) from a fixed set of possible names ("enums"). Here it's a list of regular expressions.

Also, python has nice functional aspects. So, maybe instead the new parameter should be `validate_element_fcn`, so you could do:

```python
         for (idx, v) in enumerate(value):
             try:
+                if self.validate_element_fcn is not None and not self.validate_element_fcn(v):
+                    raise ValidationError(
+                        name,
+                        "'%s' is not valid"
+                        % (value),
+                    )
                 vv = self.nested._validate(v, name + "[" + str(idx) + "]")
             except ValidationError as e:
                 raise ValidationError(e.name, e.error_message)
             result.append(vv)
```

And document `validate_element_fcn()` may either return False or raise ValidationError on error.

And you use it like:

```python
+                ArgValidatorList(
+                    "dns_options",
+                    nested=ArgValidatorStr("dns_options[?]"),
+                    validator_element_fcn= lambda v: any(re.match(patter, v) for pattern in ArgValidator_DictIP.VALID_OPTIONS),
+                    default_value=list,
+                ),
```

Reply:
actually, `ArgValidatorList` already has something like `validate_element_fcn`: that's `self.nested`. So if you want to validate the elements, set the appropriate nested function. Meaning, the right solution is not
```python
                ArgValidatorList(
                    "dns_options",
                    nested=ArgValidatorStr("dns_options[?]"),
                    enum_values=ArgValidator_DictIP.VALID_OPTIONS,
                    default_value=list,
                ),
```
but
```python
                ArgValidatorList(
                    "dns_options",
                    nested=ArgValidatorIPDnsOptions("dns_options[?]"),
                    default_value=list,
                ),

```
or,
```python
                ArgValidatorList(
                    "dns_options",
                    nested=ArgValidatorStr("dns_options[?]", regex_list=[...]),
                    default_value=list,
                ),

```

Reply:
I am not very convinced why we should not  use list of regex inside enums

Reply:
I don't understand what you mean.

`ArgValidatorList` is a validator of a some "list thingy". But the elements of the list are not necessarily strings. Hence, `ArgValidatorList` having a property with a list of regexes to validate strings seems wrong (because regex only make sense to validate strings, and the elements of `ArgValidatorList` may not be strings). I guess, you could add a subclass `ArgValidatorStringList`, but you don't need that. Instead, `ArgValidatorList` already has a validator for the list elements: `self.nested`. Set the right `self.nested` to validate the list elements with a regex.

Also, not sure what you mean with `list of regex inside enums`.

Reply:
for those lists that not containg strings, they also may not  need `enum_values` to validate their value, and right now it make more sense to include function `str()` here for those lists containing non-string type when `enum_values` is not `None` . Given the fact that python is short circuiting, I do not see any problem here for lists containing non-string type. 

the following is a list of regex inside enum_values:
```
    VALID_OPTIONS = [
        r"^attempts:([1-9]\d*|0)$",
        r"^debug$",
        r"^edns0$",
        r"^ndots:([1-9]\d*|0)$",
        r"^no-check-names$",
        r"^no-reload$",
        r"^no-tld-query$",
        r"^rotate$",
        r"^single-request$",
        r"^single-request-reopen$",
        r"^timeout:([1-9]\d*|0)$",
        r"^trust-ad$",
        r"^use-vc$",
    ]
```

Reply:
I see it this way:

With all `ArgValidator` classes you have complete control about validating the element simply by overwriting `_validate_impl()` or `_validate_post()`. That is the most flexible one, and it requires you to subclass `ArgValidator`.

But we have common behaviors, like a string. Hence, there is reusable `ArgValidatorStr` implementation to validate a string. So, if you wanted to validate `ethernet.duplex` you could do
```
class ArgValidator_EthernetDuplex(ArgValidatorStr):
     def __init__(self, name):
           ArgValidatorStr.__init__(self, "duplex", default_value=None)
     def _validate_impl(self, value, name):
           value = ArgValidatorStr._validate_impl(self, value, name)
           if value not in ["half", "full"]:
                  raise ValidateError("duplex value must be either half or full")
           return value
```
But validating against a fixed set of strings (enums!!) is a common thing. Hence, `ArgValidatorStr` helps you with such common, simple cases and you can instead do:
```
   ArgValidatorStr("duplex", default_value=None, enum_values=['half', 'full'])
```

Sidenote: the name "enum_values" here has the meaning that these strings are from a fixed set of valid strings. I don't understand why you keep insisting on naming a list of regular expressions to validate the string as "enum_values".

`enum_values` parameter exists for convenience on top of the common `_validate_impl()` code and `ArgValidatorStr._validate_impl()` implements that to be re-used.

But adding a parameter `regex` to `ArgValidatorList()` seems a bit odd to me. Why such a convenience paratmer, which hardly makes sense for the majority of list validators? Why

```
                ArgValidatorList(
                    "dns_options",
                    nested=ArgValidatorStr("dns_options[?]"),
                    regex=ArgValidator_DictIP.REGEX_DNS_OPTIONS,
                    default_value=list,
                )
```
instead of
```
                ArgValidatorList(
                    "dns_options",
                    nested=ArgValidatorStr("dns_options[?]", regex = ArgValidator_DictIP.REGEX_DNS_OPTIONS),
                    default_value=list,
                )
```
It makes sense to me that `ArgValidatorStr` has a convenience checker `regex` instead of `ArgValidatorList`.

Reply:
Yes, I like this!

There is still the `enum_values` argument in `ArgValidatorList`.


I suggested the name `regex`, because I thought maybe this could be either just one string (one "regex") or a list of regexes (like in the patch, where the list is combined with "OR"). So in the future we could do:

```python
        if self.regex is not None:
            r = self.regex
            if  isinstance(value, Util.STRING_TYPE):
                r = [r]
            if not any(re.match(x, value) for x in r):
                raise ValidationError(
                    name,
                    "is '%s' but must be one of '%s'"
                    % (value, "' '".join(sorted(self.regex))),
                )
```
and you could either do:
```python
     ArgValidatorStr(
           "dns_options[?]", regex=ArgValidator_DictIP.REGEX_DNS_OPTIONS
     )
```
or
```python
     ArgValidatorStr(
           "dns_options[?]", regex=r'^(use-vc|debug|...)$'
     )
```
with one regex only.


But that is not necessary now (because it's not used). But that is why I suggested "regex" as name and not "regex_list".

Reply:
> `enum_values` does not seem a good name for a list of regular expressions.

`choices` might be a better name for enum_values to make it's purpose clear. This could be done in a refactor PR if there is a consensus.